### ISSUE TITLE
Missing tests

### ISSUE THREAD

**Issue by Suzii** [martinbojnansky/pv247#2]
The assignment explicitely mentions, that you need to write tests for action creators, reducers and util functions. In order to be able to test your action creators (that contain side effects), you need to write proper Dependency injection. Lecture from Vit Svoboda might help you with this, or have a look at these videos: [Dependency injection](https://www.youtube.com/watch?v=0X1Ns2NRfks) and [Inversion of control](https://www.youtube.com/watch?v=-kpEP4JeEdc)

In your project you havent used and middleware for async redux action, which also makes your action creators almost untestable and it is a very bad practice to treat your redux store as a singleton that is imported from a single module everywhere. 

To summarize:
* Use dependency injection
* Install and use [redux-thunk](https://github.com/gaearon/redux-thunk) (or other async) middleware for redux. Mentioned in lecture 4 (Redux)
* Do not import store and do not call dispatch on it directly (e.g. `components/Channels.tsx -> componentWillMount, line 23` - you are already dispatching the action in the container, so why again? :) )
* **Write unit tests for reducers, action creators, utils**

**Comment by martinbojnansky** [comment]
I was able to test all reducers and api functions, however I'm struggling with thunk tests.

I'm not able to properly mock localStorage, for the purpose of testing thunk actions.
I've used library called jest-localstorage-mock, hovewer on all [api](https://github.com/martinbojnansky/PV247/blob/master/src/tests/api) tests and [src/tests/thunks/Login.test.tsx](https://github.com/martinbojnansky/PV247/blob/master/src/tests/thunks/Login.test.tsx) it works properly, until it comes to other two thunk tests [Channels](https://github.com/martinbojnansky/PV247/blob/master/src/tests/thunks/Channels.test.tsx)/[Conversation](https://github.com/martinbojnansky/PV247/blob/master/src/tests/thunks/Conversation.test.tsx). 

LocalStorage is used to get saved email in [initialState](https://github.com/martinbojnansky/PV247/blob/master/src/models/StoreState.tsx). I also tried to pass localStorage as direct dependency in constructor of initialState and tryGetUserId and it did not help.

I was not able to figure it out for a long time, can I ask for an advice?

**Comment by Suzii** [comment]
Hi,
first of all I would have strongly preferred if you used proper dependency injection everywhere (all impure functions including fetch, localStorage should have been injected if they are needed). This would greatly simplify your tests, and you would not need anything like mock fetch or local storage.

However, with respect to how much you have already done on this project, I do not want to force you to do it. But please, at least consider it on your next project :)

Anyway, after some tinkering with your codebase, I was able to get rid of 'localStorage is not defined' error. Try this:
* delete all `require('jest-localstorage-mock');` in tests (there are about 10 usages)
* create file `src/setuTests.ts` (**beware of the filepath!!!**) read more [here](https://github.com/facebook/create-react-app/blob/master/packages/react-scripts/template/README.md#srcsetuptestsjs)
* place only line in this file: `require('jest-localstorage-mock')`
* To be sure the new configuration is reflected, add `--no-cache` option to `test` command in your `package.json`: `"test": "react-scripts-ts test --env=jsdom --no-cache",`

Note that one test is still failing, probably due to some logical error.
Let me know it this helps 😉 

PS: Consider using some autoformat in your IDE on file save 😉

**Comment by martinbojnansky** [comment]
Thank you for the reply,

In the meantime I found the solution for my problem, even without proper dependency injection usage.
There was a problem, that store was created as a constant value. 
I'm a big fan of IoC & DI principles, however it seems a quite boilerplate in JS, or I'm not aware of any container/injector solution.

I've rewritten all actions to thunks and written a few thunk tests in [src/tests/thunks](https://github.com/martinbojnansky/PV247/blob/master/src/tests/thunks/) for now. Is it required to cover all of thunk actions by tests?

Thank you for your help.

**Comment by Suzii** [comment]
Glad to hear, you have fixed it 😃.
I do not require to cover all the thunks, but it would be nice, if you test at least one 'write' thunk (currently I can see there are only 'get' thunk actions tested). I expect all the implemented tests to be passing and none ignored ([`xit`](https://github.com/martinbojnansky/pv247/blob/master/src/tests/thunks/Channels.test.tsx#L28) or `xdescribe`). 
Let me know if you have any further questions 😉

**Comment by martinbojnansky** [comment]
Ignored test had been completed and I also added the test for the send message thunk action.
I've also created pre-configuration for tests as you described in previous comment.
Thank you for your help and please, let me know if anything further is required to finish the assignment.

**Comment by Suzii** [comment]
Great, I'll have a look on Monday 😉
